### PR TITLE
AA-185: EXTCODEHASH, EXTCODELENGTH, EXTCODECOPY access address with no code

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: tests/opbanning/test_op_banning.py [hunk 1]
```diff
@@ -32,6 +33,19 @@
     "SELFDESTRUCT",
 ]
 
+banned_opcodes_for_undeployed = [
+    "EXTCODESIZE",
```
Comment (drortirosh) [inline tests/opbanning/test_op_banning.py:37]: this is a rule: EXTCODESIZE_to_undeployed

File: tests/opbanning/test_op_banning.py [hunk 2]
```diff
@@ -32,6 +33,19 @@
...
+    "EXTCODEHASH",
+    "EXTCODECOPY",
```
Comment (drortirosh) [inline tests/opbanning/test_op_banning.py:39]: add rule "CALL_to_undeployed"

File: tests/opbanning/test_op_banning.py [hunk 3]
```diff
@@ -32,6 +33,19 @@
     "SELFDESTRUCT",
 ]
 
+banned_opcodes_for_undeployed = [
```
Comment (drortirosh) [inline tests/opbanning/test_op_banning.py:36]: these are "rules", not "opcodes"

File: tests/contracts/TestRulesFactory.sol [hunk 4]
```diff
@@ -54,9 +95,66 @@ contract TestRulesFactory is Stakable {
...
+                let size := 0
+                // allocate output byte array - this could also be done without assembly
+                // by using o_code = new bytes(size)
+                o_code := mload(0x40)
+                // new "memory end" including padding
+                mstore(0x40, add(o_code, and(add(add(size, 0x20), 0x1f), not(0x1f))))
+                // store length in memory
+                mstore(o_code, size)
+                // actually retrieve the code, this needs assembly
+                extcodecopy(100700, add(o_code, 0x20), 0, size)
```
Comment (drortirosh) [inline tests/contracts/TestRulesFactory.sol:136]: you don't care about the output buffer, so 
`extcodecopy(addr, 0, 0, 2)`
would be enough.
(you have "scratch storage" of 64 bytes from offset zero. That's why the "free mem memory ptr" is at mload(64)
also, your code does "extcodecopy" of size "zero". theoretically, an "assembly optimizer" can remove this opcode, which copies nothing.

File: tests/contracts/TestRulesFactory.sol [hunk 5]
```diff
@@ -24,6 +25,46 @@ contract TestRulesFactory is Stakable {
...
+        if (rule.eq("EXTCODEx_CALLx_undeployed_sender")) {
+            // CALL
+            create2address.call("");
+            // CALLCODE
+            assembly {
+                let x := mload(0x40)   // Find empty storage location using "free memory pointer"
+                mstore(x, 0xffffffff)  // Place signature at beginning of empty storage
+                mstore(add(x,0x04), 0) // Place first argument directly next to signature
+                mstore(add(x,0x24), 0) // Place second argument next to first, padded to 32 bytes
+                let res := callcode(5000, create2address, 0, x, 0x44, add(x,0x80), 0) // Issue call, providing 5k gas and 0 value to "address"
```
Comment (drortirosh) [inline tests/contracts/TestRulesFactory.sol:39]: like `call("")` above, can do here `callcode(5000,addr,0,0,0,0,0)`

File: tests/opbanning/test_op_banning.py (hunk 3 above)
Comment (forshtat) [reply tests/opbanning/test_op_banning.py]: Fixed.

File: tests/opbanning/test_op_banning.py (hunk 2 above)
Comment (forshtat) [reply tests/opbanning/test_op_banning.py]: Done.

File: tests/opbanning/test_op_banning.py (hunk 1 above)
Comment (forshtat) [reply tests/opbanning/test_op_banning.py]: Addressed.

File: tests/contracts/Create2.sol [hunk 6]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (forshtat) [reply tests/contracts/Create2.sol:1]: I need to calculate Create2 address counterfactually.

File: tests/contracts/TestRulesAccount.sol [hunk 7]
```diff
@@ -107,6 +107,24 @@ contract TestRulesAccount is IAccount, IPaymaster, Stakable {
...
+        else if (eq(rule, "DELEGATECALL_undeployed_contract")) { address(100300).delegatecall(""); return 0; }
+        else if (eq(rule, "STATICCALL_undeployed_contract")) { address(100400).staticcall(""); return 0; }
+        else if (eq(rule, "EXTCODESIZE_undeployed_contract")) return address(100500).code.length;
+        else if (eq(rule, "EXTCODEHASH_undeployed_contract")) return uint256(address(100600).codehash);
+        else if (eq(rule, "EXTCODECOPY_undeployed_contract")) {
+            bytes memory o_code;
+            assembly {
+                extcodecopy(100700, 0, 0, 2)
+            }
+            return o_code.length;
```
Comment (drortirosh) [inline tests/contracts/TestRulesAccount.sol:126]: remove o_cde completely. just return szero.

**Review by drortirosh — APPROVED** [review]

**Pull request merged**